Java 中的异常可以通过关键字进行控制，主要有 try、catch、throw、throws和finally：
+ try：使用try将可能出现异常的代码段包围起来，如果 try 语句块中出现异常，那么这个异常就会被抛出
+ catch：捕获异常，并在这个语句块中对该异常进行处理，如输出说明语句等
+ finally：不管发生还是不发生异常，都需要执行
+ throw：用来手动引发一个异常
+ throws：声明一个方法会抛出异常

处理异常的一般形式：
```
public class AbnormalTest{
    try{
        // 要检查是否有异常的代码段
    } catch(Exceptin e1){
        System.out.println("该处发生 e1 异常");
    } catch(Exception e2){
        System.out.println("该处发生 e2 异常");
    } finally{
        // 必须要执行的语句
    }
}
```

异常的层次结构：
+ Throwable：所有异常和错误的父类
    + Error：定义 Java 运行时的内部错误以及资源耗尽错误。表示很难恢复的错误，如内存越界。一般不期望用户来处理。
    + Exception：所有异常的父类，程序设计不合理所导致，Java 认为可以恢复的。
        + IOException：因环境因素映影响，查找不到文件，不完整路径
        + RuntimeException：用来表示设计或者实现方面的问题，如数组越界、类型转换错误、去访问一个没有初始化的对象。在出现这个错误的时候，一般不需要处理，运行的时候，环境会输出一条提示语句，提示用户修正错误。
    

uncheck Exception：Error子类和 RuntimException子类。
checked Exception：其他异常
未检查异常：
+ 类型异常转换
+ 数组访问越界
+ 空指针访问

已检查异常：
+ 在文件尾部之后读取
+ 打开一个不存在的链接
+ 反射机制中查找一个不存在的类

异常发生，Java 将该异常包装成 异常类对象，并将其引用传递给相应的 catch 语句，catch语句对异常对象进行操作。

自定义异常类
```java
class 类名 extends Exception{
    // 类体
}
```

